pkg/policy/types/entry.go

```diff
@@ -0,0 +1,245 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package types
+
+import "strconv"
+
+type Precedence uint32
+
+const (
+	PrecedenceDeny              Precedence = 1 << 7
+	PrecedenceProxyPriorityMask Precedence = PrecedenceDeny - 1 // 0-127
+	MaxListenerPriority                    = 100
+	MaxPrecedence                          = ^Precedence(0)
+	MaxDenyPrecedence                      = MaxPrecedence
+	MaxAllowPrecedence                     = MaxPrecedence & ^(PrecedenceDeny | PrecedenceProxyPriorityMask)
+)
+
+// MapStateEntry is the configuration associated with a Key in a
+// MapState. This is a minimized version of policymap.PolicyEntry.
+type MapStateEntry struct {
+	// Precedence encodes the relative order in which policy entries are selected
+	// Higher values have higher precedence.
+	// Deny and Listener priority are encoded into the precedence field.
+	Precedence Precedence
+
+	// The proxy port, in host byte order.
+	// If 0 (default), there is no proxy redirection for the corresponding
+	// Key. Any other value signifies proxy redirection.
+	ProxyPort uint16
+
+	// Invalid is only set to mark the current entry for update when syncing entries to datapath
+	Invalid bool
+
+	// AuthRequirement is non-zero when authentication is required for the traffic to be
+	// allowed, except for when it explicitly defines authentication is not required.
+	AuthRequirement AuthRequirement
+}
+
+// String returns a string representation of the MapStateEntry
+func (e MapStateEntry) String() string {
+	var authText string
+	if e.AuthRequirement != 0 {
+		var authNote string
+		if !e.AuthRequirement.IsExplicit() {
+			authNote = " (derived)"
+		}
+		authText = ",AuthType=" + e.AuthRequirement.AuthType().String() + authNote
+	}
+
+	return "Precedence=" + strconv.FormatUint(uint64(e.Precedence), 10) +
+		",ProxyPort=" + strconv.FormatUint(uint64(e.ProxyPort), 10) +
+		",IsDeny=" + strconv.FormatBool(e.IsDeny()) +
+		authText
+}
+
+// NewMapStateEntry creeates a new MapStateEntry
+// 'deny' is encoded into the PrecedenceDeny bit
+// Listener 'priority' is encoded in to the low 7 bits of 'precedence', inverted
+func NewMapStateEntry(deny bool, proxyPort uint16, priority uint8, authReq AuthRequirement) MapStateEntry {
+	if deny && (proxyPort != 0 || priority != 0 || authReq != 0) {
+		panic("Invalid parameters for NewMapStateEntry")
```

Comment:
`panic` makes me incredibly nervous. I understand that attempting to configure deny and some non-deny feature at the same time doesn't make any sense, but putting panics into production code can end up having severe impacts and I'm not entirely confident that we are running tests with all permutations to ensure this is never triggered.

Deny takes precedence anyway, so if you're trying to avoid error checking we could alternatively do something like the following example (which should also be caught by CI given we detect & fail on warnings now).

```suggestion
		log.WithFields(...).Warning("BUG: Invalid parameters for NewMapStateEntry. Please report this bug to Cilium developers.")
		proxyPort = 0
		priority = 0
		authReq = 0
```

Reply:
I guess current examples also use `hclog.Stacktrace()`, you could consider that too.

Reply:
Replaced the panic with normalizing the parameters (without any logging). This is now mature enough for that :-)